btree/btree.c

```diff
@@ -2035,11 +2191,877 @@ static void generic_move(struct nd *src, struct nd *tgt,
+	struct m0_btree_oimpl *oi         = bop->bo_i;
+	struct level          *curr_level;
+
+	if (oi)
+		curr_level = &oi->i_level[oi->i_used];
+
+	switch (bop->bo_op.o_sm.sm_state) {
+	case P_INIT:
+		if ((flags & OF_COOKIE) &&
+		    cookie_is_set(&bop->bo_rec.r_key.k_cookie))
+			return P_COOKIE;
+		else
+			return P_SETUP;
+	case P_COOKIE: {
+		void *cookie_segaddr = bop->bo_rec.r_key.k_cookie.segaddr;
+		if (cookie_is_valid(tree, &bop->bo_rec.r_key.k_cookie) &&
+		    !cookie_overflow_is_possible(cookie_segaddr))
+			return P_LOCK;
+		else
+			return P_SETUP;
+	}
+	case P_SETUP: {
+		bop->bo_arbor->t_height = tree->t_height;
+		bop->bo_i = alloc(bop->bo_arbor->t_height);
+		if (bop->bo_i == NULL)
+			return fail(bop, M0_ERR(-ENOMEM));
+		return P_LOCKALL;
+	}
+	case P_LOCKALL:
+		if (bop->bo_flags & OF_LOCKALL) {
+			//return m0_sm_op_sub(&bop->bo_op, P_LOCK, P_DOWN);
+			m0_rwlock_write_lock(&tree->t_lock);
+		}
+		/* Fall through to the next stage */
+	case P_DOWN:
+		oi->i_used = 0;
+		/* Load root node. */
+		return node_get(&oi->i_nop, tree, &tree->t_root->n_addr,
+				P_NEXTDOWN);
+	case P_NEXTDOWN:
+		if (oi->i_nop.no_op.o_sm.sm_rc == 0) {
+			struct slot    node_slot = {};
+			struct segaddr child_node_addr;
+
+			curr_level->l_node = oi->i_nop.no_node;
+			node_slot.s_node = oi->i_nop.no_node;
+			curr_level->l_seq = curr_level->l_node->n_seq;
+			/* verify node footer */
+			/*int rc = node_verify(curr_level->l_node);
+			if (rc)
+			{
+				node_op_fini(&oi->i_nop);
+				return fail(bop, rc);
+			}*/
+			oi->i_nop.no_node = NULL;
+
+			node_find(&node_slot, &bop->bo_rec.r_key);
+			curr_level->l_idx = node_slot.s_idx;
+			if (node_level(node_slot.s_node) > 0) {
+				node_child(&node_slot, &child_node_addr);
+				/* TBD: check the address of child node is
+				within segaddr boundaries */
+				int rc = address_is_valid(child_node_addr);
+				if (rc) {
+					node_op_fini(&oi->i_nop);
+					return fail(bop, rc);
+				}
+				oi->i_used++;
+				return node_get(&oi->i_nop, tree,
+						&child_node_addr, P_NEXTDOWN);
+			} else
+				return P_ALLOC;
+		} else {
+			node_op_fini(&oi->i_nop);
+			return fail(bop, oi->i_nop.no_op.o_sm.sm_rc);
+		}
+	case P_ALLOC: {
+		/* validate curr_level->l_node(i.e.is it still exists or not) */
+		int rc = is_node_valid(curr_level->l_node);
+		if (rc) {
+			node_op_fini(&oi->i_nop);
+			return fail(bop, rc);
+		}
+		if (m0_btree_overflow_is_possible(curr_level->l_node))
+			return m0_btree_put_alloc_phase(bop);
+
+		/* reset oi->i_used */
+		oi->i_used = bop->bo_arbor->t_height - 1;
+		return P_LOCK;
+	}
+	case P_LOCK:
+		if (!locked(tree))
+			return lock_op_init(&bop->bo_op, &bop->bo_i->i_nop,
+					    P_CHECK);
+		else
+			return P_CHECK;
+	case P_CHECK: {
+		oi->i_trial++;
+		if (!path_check(oi, tree, &bop->bo_rec.r_key.k_cookie)) {
+			if (oi->i_trial == MAX_TRIALS) {
+				if (bop->bo_flags & OF_LOCKALL)
+					return fail(bop, -ETOOMANYREFS);
+				else
+					bop->bo_flags |= OF_LOCKALL;
```

Comment:
Here don't we want to restart the state machine instead of fall down ?

Reply:
I am not sure but , according to height decrease or increase it will have to go to next state right?

Reply:
Yes, you are right. Height check should take care of moving state machine to correct state.